feat: add configuration to audit log to a rotating file (#1443)
janus-idp/backstage-showcase: showcase-docs/audit-log.md

# Audit Logging for Backstage Showcase

> Backstage Showcase currently supports audit logging for the catalog and scaffolder. Audit logs are signified by the `isAuditLog: true` field.

Reviewer: ```suggestion
Backstage Showcase supports audit logging for both the catalog and scaffolder. Audit logs can be identified by the isAuditLog: true field.
```

## Configuring the audit logger

> ### Configuring the audit logging to console

Reviewer: ```suggestion
### Logging to Console
```

> By default, the audit logger will log to the console. You can disable this behavior by setting the following in your `app-config.yaml`:

Reviewer: ```suggestion
By default, the audit logger logs to the console. To disable this, update your app-config.yaml with:
```

```yaml
auditLog:
  logToConsole: false
```

> ### Configuring the audit logging to a rotating file

Reviewer: ```suggestion
### Logging to a Rotating File
```

> #### Enabling Audit Logging to a rotating file

Reviewer: ```suggestion
#### Enabling Rotating File Logging
```

> Backstage Showcase also supports forwarding the audit logs to a rotating file. By default this is disabled, to enable it set the following:

Reviewer: ```suggestion
To enable audit logging to a rotating file, set the following in your configuration (this feature is disabled by default):
```

```yaml
auditLog:
  rotate:
    enabled: true
```

> The default behavior with the above configuration is:
> 
> - Rotate at midnight of the local system timezone
> - Output logs to a file with the format of: `redhat-developer-hub-audit-%DATE%.log`
> - Store log files in the `/var/log/redhat-developer-hub/audit` directory
> - Does not remove old logs
> - Does not gzip archived logs
> - No file size limit

Reviewer: ```suggestion
With this configuration, the default behavior is:

- Rotate logs at midnight (local system timezone)
- Log file format: redhat-developer-hub-audit-%DATE%.log
- Log files stored in /var/log/redhat-developer-hub/audit
- No automatic log file deletion
- No gzip compression of archived logs
- No file size limit
```

> #### Configuring custom location and file name

Reviewer: ```suggestion
#### Customizing Log File Location and Name
```

> By default, the audit logs will be written in the `/var/log/redhat-developer-hub/audit` directory. To customize the destination in which the audit log files will be generated, provide the path to the directory (an absolute path is recommended):

Reviewer: ```suggestion
To change the directory where log files are stored, specify a custom path (an absolute path is recommended): By default, the audit logs are written in the `/var/log/redhat-developer-hub/audit` directory.
```

```yaml
auditLog:
  rotate:
    logFileDirPath: /custom-path
```

---

**NOTE**

> If no directory exists in the specified path, the directory will be automatically generated

Reviewer: ```suggestion
The specified directory will be created automatically if it does not exist.
```

---

By default, the audit log files will be in the following format: `redhat-developer-hub-audit-%DATE%.log` where `%DATE%` is the format specified in [`auditLog.rotate.dateFormat`](#configuring-the-file-rotation-frequency).

> You can customize it by setting your custom file format:

Reviewer: ```suggestion
To customize the log file name format, use:
```

```yaml
auditLog:
  rotate:
    logFileName: custom-audit-log-%DATE%.log
```

> #### Configuring the file rotation frequency

Reviewer: ```suggestion
#### Configuring File Rotation Frequency
```

> By default, the file rotation would occur daily at 00:00 of the local timezone of the system the backstage showcase is running on. There are multiple configurations to configure the file rotation frequency:

Reviewer: ```suggestion
The default file rotation occurs daily at 00:00 local time. You can adjust the rotation frequency with the following configurations:
```

```yaml
auditLog:
  rotate:
    frequency: '12h' # Default: `custom`
    dateFormat: 'YYYY-MM-DD' # Default: `YYYY-MM-DD`
    utc: false # Default: `false`
    maxSize: 100m # Default: undefined
```

> The `frequency` configuration supports the following inputs:
> 
> - `daily`: Rotates daily at 00:00 of the local timezone
> - `Xm`: Rotates every X minutes
> - `Xh`: Rotates every X hours
> - `test`: Rotates every 1 minute
> - `custom`: Uses the `dateFormat` to configure the rotation frequency. This is the default configuration if `frequency` is not provided.

Reviewer: ```suggestion
`frequency` options include:

- `daily`: Rotate daily at 00:00 local time
- `Xm`: Rotate every X minutes
- `Xh`: Rotate every X hours
- `test`: Rotate every 1 minute
- `custom`: Use `dateFormat` to set the rotation frequency (default if frequency is not specified)
```

> The `dateFormat` configuration configures both the `%DATE%` value used by the `logFileName`, as well as the file rotation frequency if `frequency` is set to `custom`. The default format is `YYYY-MM-DD` meaning it will rotate the file whenever the `YYYY-MM-DD` changes (daily rotate). The supported values are those used by [Moment.js](https://momentjs.com/docs/#/displaying/format/). The file will rotate whenever the `%DATE%` in the provided format changes if

Reviewer: ```suggestion
The dateFormat setting configures both the %DATE% in logFileName and the file rotation frequency if frequency is custom. The default format is YYYY-MM-DD, meaning daily rotation. Supported values are based on [Moment.js formats](https://momentjs.com/docs/#/displaying/format/).
```

Examples:

```yaml
auditLog:
  rotate:
    # If you want rotations to occur every week for some reason. Example `%DATE$` = '2025-Jul-Week 30'
    dateFormat: 'YYYY-MMM-[Week] ww'
```

```yaml
auditLog:
  rotate:
    # If you want to rotate at noon and midnight
    dateFormat: 'YYYY-MM-DD-A'
```

> By default, the `dateFormat` being used is in the `local` time of the system timezone. If you want to use `utc`, set the following:

Reviewer: ```suggestion
To use UTC time instead of local time:
```

```yaml
auditLog:
  rotate:
    utc: true # Default: False
```

> We can also set a `maxSize` audit logs can be before a rotation is automatically triggered. The file rotation in this case will add a count (1,2,3,4,...) at the end of the filename when the required size is met. Ex: `redhat-developer-hub-audit-2024-07-22.log.3`.
> 
> By default no `maxSize` is configured. To configure `maxSize`, provide a number followed by one of `k`, `m`, or `g` to specify the file size in kilobytes, megabytes, or gigabytes.

Reviewer: ```suggestion
To set a maximum log file size before rotation (adding a count suffix to the filename upon reaching the limit): Ex: `redhat-developer-hub-audit-2024-07-22.log.3`. 

To configure `maxSize`, provide a number followed by one of `k`, `m`, or `g` to specify the file size in kilobytes, megabytes, or gigabytes.
```

```yaml
auditLog:
  rotate:
    maxSize: 100m
```

> #### Configuring the file retention policy

Reviewer: ```suggestion
#### Configuring File Retention Policy
```

> By default, no files are deleted and no files are archived.
> 
> You can configure the max number of files to keep before log deletion of the oldest log begins by configuring `maxFilesOrDays` with a number:

Reviewer: ```suggestion
By default, log files are not deleted or archived. You can configure the maximum number of files to keep:
```

```yaml
auditLog:
  rotate:
    maxFilesOrDays: 14 # Deletes the oldest log when there are more than 14 log files
```

> You can also configure the max number of DAYS to keep files around by appending a `d` after the number. Note: this is mutually exclusive with the number of files configuration:

Reviewer: ```suggestion
Or, configure the maximum number of days to retain logs:
```

```yaml
auditLog:
  rotate:
    maxFilesOrDays: 5d # Deletes logs older than 5 days
```

> You can also configure the audit logger to archive and compress rotated audit logs using `gzip` to save space:

Reviewer: ```suggestion
To archive and compress rotated logs using gzip:
```
